### DISCUSSION TITLE
Quality of life improvements

### DISCUSSION THREAD

**Discussion by fuegovic** [danny-avila/LibreChat#89] [Feature Requests & Suggestions]
After spending some time using the chatgpt-clone I compiled a list of quality of life improvements that could benefit to the project. 

Those are only suggestions based on my experience. I tried to mark all the features already suggested or planned with a " * "

---
- Editing a "CustomGPT" model should open the window to allow to edit the "Prompt Prefix" as well as the "Custom Name"
- Regarding the Bing censorship, it would be helpful to be able leave the partially generated message even if Bing removes it. (maybe in a different color or with an annotation that makes it clear that it ha been censored and is not part of the chat
- <del>The logout button seems to be useless, it could be removed</del> (my mistake, there's a use for it)
- When a previous chat is selected it should automatically scroll the bottom of the page
- Get an approximate tokens amount used for the chat or answer (This message was generated using 18 tokens ~= $0.000036) 
	- see https://github.com/Niek/chatgpt-web for an example of the implementation
- <del>Being able to stops/cancel the generation of an answer</del> (recently added)
- Add the option to export a conversation (txt file)
-  When using bing, show the number of interaction left (like on bing's website)
- A scrollable list of the conversation history instead of multiple pages
- *(Planned) Clear conversations should ask for confirmation before executing to prevent accidents
- *(Suggested/planned) Having an easy access to the "temperature" setting of chatGPT (creativity/randomness level)
- *(Suggested) Have multiple model choices for chatGPT (gpt-4, gpt-3.5-turbo, text-davinci-003, code-davinci-002...)

- Having access to a basic settings menu where:
	- you have the option to access through a public url
	- you can add/remove models from the default list
	- you are able to easily customize the user name
	- **(suggested/planned) you can select more choices than Sydney for Bing's jailbreak, and have some basic CustomGPT model there (* Similar to the "Prompt Templates" planned feature)

Optimistic	
- Simplify the install process using scripts (similar to automatic1111)

**Comment by danny-avila** [comment]
Great suggestions. I've got some thoughts, been a busy day, but I'll get back with a reply on each point

**Reply by fuegovic** [reply]
Thanks, don't worry about it, it's still the best implementation I found so far!

**Comment by wtlyu** [comment]
logout buttom works when you enable the usersystem. it will logout for your account and then request login. 

the usersystem conyains only a sample endpoint which will automaticly log you in to Sample User. So you will come back to home page and seems nothing happened.

**Reply by fuegovic** [reply]
my mistake, thanks for the clarification!

**Comment by wtlyu** [comment]
Being able to stops/cancel the generation of an answer

it's also added yesterday.

**Reply by fuegovic** [reply]
Great, I noticed after updating on my side, I updated my first post to show that it is now implemented

**Comment by wtlyu** [comment]
I'm working on s total structure reactor on this project. some of the features looks pretty good for me.

**Comment by danny-avila** [comment]
First off, thanks for taking the time to write all this. Glad you're enjoying it. Anything i'm noting here I'm putting on my personal list to touch up on.

>     * Editing a "CustomGPT" model should open the window to allow to edit the "Prompt Prefix" as well as the "Custom Name"

Noted! This is definitely an improvement over what it is now

>     * Regarding the Bing censorship, it would be helpful to be able leave the partially generated message even if Bing removes it. (maybe in a different color or with an annotation that makes it clear that it ha been censored and is not part of the chat
> 

I'm not getting any responses before censorship, perhaps this is only seen on sydney which is unavailable (not yet up to date with latest jailbreak)

>     * When a previous chat is selected it should automatically scroll the bottom of the page
>     * Get an approximate tokens amount used for the chat or answer (This message was generated using 18 tokens ~= $0.000036)

* These seem like preference items. Specifically, scrolling to bottom was the default behavior at first and i didnt like it. I would like to make this part of a configurable system to hide and show things as well, which would be features build on top some of the things also mentioned below.

>     * Add the option to export a conversation (txt file)

noted: Should be easy enough!


>     * When using bing, show the number of interaction left (like on bing's website)

Planned (although I've seen this number change several times throughout this project)

>     * A scrollable list of the conversation history instead of multiple pages

This was the original design and I could bring it back as a preference item too. I prefer this too.

>     * Having access to a basic settings menu where:
>       
>       * you have the option to access through a public url
>       * you can add/remove models from the default list
>       * you are able to easily customize the user name

See bullet point. im adding these specifically to the roadmap. technically you can remove models by not providing the credentials.

> Optimistic
> 
>     * Simplify the install process using scripts (similar to automatic1111)

Noted!


Last but not least, I would like to mention these features can be accelerated by community support. I have a couple planned features to go but then i will have to pause my own active development on more until there is a proper testing environment. I could definitely use the most help on this if nothing else.

**Reply by fuegovic** [reply]
I appreciate you taking the time to consider my suggestions. While I'd like to be able to contribute more than ideas I'm afraid I'd do more harm than good considering my coding skills.

Regarding the Bing censorship, it's something that happened to me quite a lot when trying to use Bing for creativity. This problem arises particularly in longer texts such as song lyrics or short stories. To replicate this issue, I have discovered that asking Bing for a "short detective story with a dark and mysterious subject" is quite effective.

Bing will then proceed to compose most of the story until it describe a murder scene or something slightly graphic. At that point Bing suddenly remove everything replacing it with an apology message instead. The same problem is present in the official Bing chat interface. This flaw can be frustrating when trying to use Bing for anything creative.

I've had this issue with other subject but this is the most efective way I found to replicate the issue.


https://user-images.githubusercontent.com/32828263/226444299-3f87852d-efc3-445b-8101-cbde07c27c88.mp4

**Reply by danny-avila** [reply]
thanks for that example. i think this will be alleviated with the fix for sydney. im testing it now and managed to get 1 complete response for that prompt

![image](https://user-images.githubusercontent.com/110412045/226465508-5ecfaaa5-f275-4dc1-aaac-ac4106089e10.png)

**Reply by danny-avila** [reply]
ill for sure work on handling the sudden stop, definitely can

**Reply by danny-avila** [reply]
with long responses you have to keep prodding it to continue 
![image](https://user-images.githubusercontent.com/110412045/226467605-f234573d-0b52-421f-96cb-f6488c30bad9.png)

**Reply by danny-avila** [reply]
change is live!

**Comment by fuegovic** [comment]
I just pulled the update from the "main" branch and encountered the same behavior. Note that this happens mostly with bing, I already noticed that Sydney is way less prone to this kind of censorship.

I guess that what I'd like would implies to log the prompt during its generation and having the script look for a switch of the prompt "contentOrigin" from 'DeepLeo' to 'Apology' mid prompt and then append the "refusal to answer" from bing to the partially generated and deleted answer to make it clear that the answer was censored by Bing...

That seems complicated, I could also start to use Sydney for those kind of prompts 😉

https://user-images.githubusercontent.com/32828263/226495235-682187ec-6e20-432b-b044-7a76a92dd1c5.mp4

**Comment by alfredo-f** [comment]
```
* Get an approximate tokens amount used for the chat or answer (This message was generated using 18 tokens ~= $0.000036)
```

I couldn't understand if this is planned.

**Reply by fuegovic** [reply]
I do not think it is planned, this list is not up to date with the implemented and planned features anymore... You can check out the roadmap section of the readme or the discord to get information that is more up to date with the progress!

**Reply by danny-avila** [reply]
The functionality to approximate token usage before the message is sent is already there, but not being used. I'll look into including the total tokens used as an optional message ouput from the api response, along with before the message is sent
